crucible-mir/src/Mir/Intrinsics.hs

```diff
@@ -801,6 +804,121 @@ readMirAggregateWithSymOffset bak iteFn off tpr ag@(MirAggregate totalSize m)
+    offsetIn ::
+      NonEmpty Word ->
+      MuxLeafT sym IO (SymExpr sym BaseBoolType)
+    offsetIn (off0 :| offs) = liftIO $ do
+      -- We want to construct a symbolic expression representing whether or not
+      -- `off` appears in this aggregate's candidate offsets. The most
+      -- straightforward way to do that is to test it for equality against each
+      -- of those offsets and `orPred` those tests into a single expression.
+      -- However, expressions like that can be expensive to simulate.
+      --
+      -- If portions of the aggregate represent arrays, then those entries will
+      -- be evenly spaced a known stride apart from one another. If we can find
+      -- such contiguous "runs" of entries, we can construct much simpler
+      -- membership predicates for them: is the provided offset between the
+      -- beginning and the end of a run, and is the offset a multiple of the
+      -- stride of that run?
+      --
+      -- We partition the aggregate into contiguous runs and discrete elements
+      -- to take advantage of the more efficient predicate where possible, and
+      -- to fall back to the less efficient predicate where necessary.
+      let spans = List.zipWith (,) (off0 : offs) (offs <> [totalSize])
+          (offsets, runs) = foldRuns spans
+
+      offsetsPred <- orPredBy (\o -> bvEq sym off =<< offsetLit o) offsets
+      runsPred <- orPredBy runPred runs
+
+      orPred sym runsPred offsetsPred
+
+    orPredBy :: (a -> IO (Pred sym)) -> [a] -> IO (Pred sym)
+    orPredBy f xs = do
+      case xs of
+        [] -> pure (falsePred sym)
+        (z : zs) -> do
+          zPred <- f z
+          zsPred <- mapM f zs
+          foldM (orPred sym) zPred zsPred
+
+    -- Whether `off` appears in the given `Run` of aggregate elements.
+    runPred :: Run -> IO (Pred sym)
+    runPred run = do
+      -- Note that we're using the unique stride as a proxy for element type
+      -- width, since we currently hardcode all aggregate entries as having
+      -- width 1 (TODO: hardcoded size=1). Once we move beyond that hardcoding,
+      -- we should use real entry widths here, rather than a proxy.
+      let tyWidth = rStride run
+
+      -- off >= rFrom
+      afterBeginning <- bvUge sym off =<< wordLit sym (rFrom run)
+
+      -- off < rTo (not `<=` because `rTo` is exclusive)
+      beforeEnd <- bvUlt sym off =<< wordLit sym (rTo run)
+
+      inBounds <- andPred sym afterBeginning beforeEnd
+
+      -- (off - rFrom) `mod` tyWidth == 0
+      relativeOff <- bvSub sym off =<< wordLit sym (rFrom run)
+      offModWidth <- bvUrem sym relativeOff =<< wordLit sym tyWidth
+      atTyBoundary <- bvEq sym offModWidth =<< wordLit sym 0
+
+      andPred sym inBounds atTyBoundary
+
+    -- Given a sorted list of element "spans", represented as @(fromOffset,
+    -- toOffset)@ pairs, find and return all contiguous sequences of two or more
+    -- elements of the same width, and the offsets all other elements.
```

Comment:
Does the range `(fromOffset, toOffset)` have inclusive bounds? Exclusive? Worth clarifying here.

Reply:
`(0, 2)` would refer to a value whose byte width is two, so whose bytes appear at 0 and 1 - so, left-inclusive, right-exclusive. I've clarified this in ac1876c4e.